RealmSwift/Tests/RealmCollectionTypeTests.swift

```diff
@@ -522,6 +522,66 @@ class RealmCollectionTypeTests: TestCase {
         XCTAssertTrue(collection.freeze().isFrozen)
     }
 
+    func testThaw() {
+        let frozen = collection.freeze()
+        XCTAssertTrue(frozen.isFrozen)
+
+        let frozenRealm = frozen.realm!
+        assertThrows(try! frozenRealm.write {}, reason: "Can't perform transactions on a frozen Realm")
+
+        let live = frozen.thaw()
+        XCTAssertFalse(live.isFrozen)
+
+        let liveRealm = live.realm!
+        try! liveRealm.write { liveRealm.delete(live) }
+        XCTAssertTrue(live.isEmpty)
+        XCTAssertFalse(frozen.isEmpty)
+    }
+    
+    func testThawFromDifferentThread() {
+        let frozen = collection.freeze()
+        XCTAssertTrue(frozen.isFrozen)
+
+        dispatchSyncNewThread {
+            let live = frozen.thaw()
+            XCTAssertFalse(live.isFrozen)
+
+            let liveRealm = live.realm!
+            try! liveRealm.write { liveRealm.delete(live) }
+            XCTAssertTrue(live.isEmpty)
+            XCTAssertFalse(frozen.isEmpty)
+        }
+    }
+    
+    func testThawPreviousVersion() {
+        let frozen = collection.freeze()
+        XCTAssertTrue(frozen.isFrozen)
+        XCTAssertEqual(collection.count, frozen.count)
+        
+        let realm = collection.realm!
+        try! realm.write({ realm.delete(collection) })
+        XCTAssertNotEqual(frozen.count, collection.count, "Frozen collections should not change")
+        
+        let live = frozen.thaw()
+        XCTAssertTrue(live.isEmpty, "Thawed collection should reflect transactions since the original reference was frozen")
+        XCTAssertFalse(frozen.isEmpty)
+        XCTAssertEqual(live.count, self.collection.count)
+    }
+    
+    func testThawDeletedParent() {
```

Comment:
@tgoyne 
>That at a version where the object or list's parent object has been deleted

A couple questions on this:
This is technically using the Object.thaw inside of the Collection tests. But the collection tests are already well suited to running through every collection type. Should this be in Object tests?

When thawing an object whose parent link was deleted creates an invalid object. What _should_ a developer see if they hit this case? Nil? An exception?

<img width="1401" alt="Screen Shot 2020-12-17 at 6 16 26 PM" src="https://user-images.githubusercontent.com/25092858/102556118-2716a680-4096-11eb-9149-ad3b02e6fcb1.png">

Reply:
thaw() should probably return nil if the object no longer exists.

Reply:
This would mean that thaw has to be tried, right?

Reply:
Why wouldn't this be true for frozen as well?

Reply:
freeze() always returns an object at the same version as the source object, and you can't call it on an object that doesn't exist. thaw() may give you an object at a different version if there's already a Realm open on the current thread.